alembic/versions/2023_09_12_1726-93d7c3c93a06_code_provenance_pk.py

```diff
@@ -0,0 +1,103 @@
+"""code_provenance_pk
+
+Revision ID: 93d7c3c93a06
+Revises: 04e5cdfa1ad9
+Create Date: 2023-09-12 17:59:07.897131
+
+"""
+from alembic import op
+import sqlalchemy as sa
+
+
+# revision identifiers, used by Alembic.
+revision = '93d7c3c93a06'
+down_revision = '04e5cdfa1ad9'
+branch_labels = None
+depends_on = None
+
+
+def upgrade() -> None:
+    # Alembic autogenerate totally screwed this up, so these were all
+    # done manually.
+    #
+    # This migration DOES NOT preserve database information!  To do
+    # that, we'd also need to write the commands that connect the ids of
+    # the current foreign keys to the versions and hashes that are
+    # replacing them.  Since we don't have any existing databases that
+    # can't be blow away right now, I'm not worrying about this.
+    op.drop_index('ix_code_hashes_hash', table_name='code_hashes')
+    op.drop_column( 'code_hashes', 'id' )
+    op.alter_column( 'code_hashes', 'hash', new_column_name='id' )
+    op.create_primary_key( 'pk_code_hashes', 'code_hashes', [ 'id' ] )
+
+    op.drop_column( 'code_hashes', 'code_version_id' )
+    op.drop_column( 'provenances', 'code_version_id' )
+    op.drop_index('ix_code_versions_version', table_name='code_versions')
+    op.drop_column( 'code_versions', 'id' )
+    op.alter_column( 'code_versions', 'version', new_column_name='id' )
+    op.create_primary_key( 'pk_code_versions', 'code_versions', [ 'id' ] )
```

Comment:
What primarily motivated this was models.base.safe_merge ; that works as is if the primary key for all models is "id".

I originally defined a class variable _pk that defined what the primary key was, and then did a getattr in safe_merge.  However, I decided that that was going to be more fiddily if we ever needed to create another class with something besides id as the primary key, and it was simpler to just keep id as the primary key for everything.

Reply:
OK, that's a good reason. I guess it is kind of a hassle to update it all to point to id but since you already took care of that, it would probably save us headaches in the future.